src/XtensaOptimize.cpp

```diff
@@ -2063,6 +2063,284 @@ class SimplifySliceConcat : public IRGraphMutator {
+// defined in the given loop. If these conditions are met then allocation can
+// be lifted out of the given loop (or potentially out of multiple of loops).
+// The difficulty is that due to loop partitioning there might be a multiple
+// instances of the same allocation in different instance of the loops. If all
+// instances can be lifted then we can have a single allocation outside of the
+// loop (we need to take max ot total extents as an extent for it to make sure
+// all instances are satisfied).
+// However, there might be a situation when only some of the allocations can be
+// lifted, which potentially could lead to duplicate allocation being generated.
+// Consider an example:
+// for yo:
+//   for xo:
+//        for x:
+//           allocate tmp[depends on let's inside of the loop x]
+//        for x:
+//           allocate tmp[doesn't depend on let's inside of the loop x]
+//        for x:
+//           allocate tmp[depends on let's inside of the loop x]
+// in which if we only consider a local context, the tmp allocation can be
+// lifted from the second inner loop, but couldn't be lifted from the first and
+// third loops.
+// In order to handle cases like this, LiftAllocations has two passes:
+// ** During the first pass, for each loop we compute a list of allocations which
+// are *not* allowed to be lifted to this loop level. This is done by finding a
+// *last* loop (i.e we are not able to lift allocation from the parent loop)
+// from which we *can* lift the allocation and then updating all of it's parent
+// loop lists to disallow given allocation.
+// ** In the second pass, we simply push allocations as far as we can up the
+// loop nest, while making sure that they satisfy all of the conditions (
+// including absence in the "disallowed" list).
+class LiftAllocations : public IRMutator {
+    using IRMutator::visit;
+
+    // Index of the current loop.
+    int loop_index = 0;
+    // Is this a first sub-pass of the pass?
+    bool first_run = true;
+
+    struct Allocation {
+        std::string name;
+        Type type;
+        MemoryType memory_type;
+        Expr total_extent;
+        int const_allocation_size;
+
+        Expr condition;
+        Expr new_expr;
+        std::string free_function;
+        int padding;
+
+        Allocation(std::string name, Type type, MemoryType memory_type,
+                   Expr total_extent, int const_allocation_size,
+                   Expr condition, Expr new_expr,
+                   std::string free_function, int padding)
+            : name(std::move(name)),
+              type(type),
+              memory_type(memory_type),
+              total_extent(std::move(total_extent)),
+              const_allocation_size(const_allocation_size),
+              condition(std::move(condition)),
+              new_expr(std::move(new_expr)),
+              free_function(std::move(free_function)),
+              padding(padding) {
+        }
+    };
+
+    struct Loop {
```

Comment:
Same comment as above, but with For nodes instead.

Reply:
Done